broker/src/tunneldigger_broker/broker.py

```diff
@@ -77,6 +83,26 @@ def create_tunnel(self, broker, address, uuid, remote_tunnel_id, client_features
         else:
             tunnel_str = "(%s)" % uuid
 
+        # Rate limit creation of new tunnels with the same IP address.
+        # Runs broker.tunnel-connection-threshold hook if threshold exceeded:
+        if self.connection_rate_limit_per_ip_count > 0 and self.connection_rate_limit_per_ip_time > 0:
+            if address[0] not in self.last_tunnel_created_per_ip:
+                self.last_tunnel_created_per_ip[address[0]] = deque([now]) # Create deque collection if IP address does not exist in dict
+            if len(self.last_tunnel_created_per_ip[address[0]])-1 is self.connection_rate_limit_per_ip_count:
+                scale_delta = now - self.last_tunnel_created_per_ip[address[0]][0] # Delta of oldest timestamp in collection and now
+                if scale_delta <= self.connection_rate_limit_per_ip_time:
+                    logger.info("Tunnel %s connection threshold exceeded (%s in %s (%s) seconds). Running hook" % (tunnel_str, self.connection_rate_limit_per_ip_count, int(scale_delta), self.connection_rate_limit_per_ip_time))
+                    broker.hook_manager.run_hook(
+                        'broker.tunnel-connection-threshold',
+                        address[0],
+                        address[1],
+                        uuid,
+                    )
+                    del self.last_tunnel_created_per_ip[address[0]]
+                    return False
+                self.last_tunnel_created_per_ip[address[0]].popleft() # Delete first (oldest) timestamp in deque collection
+            self.last_tunnel_created_per_ip[address[0]].append(now) # Append current timestamp at the end of deque collection
```

Comment:
At this point, maybe assert that `len(self.last_tunnel_created_per_ip[address[0]])` is not too big? Your `is` above would go horribly wrong if the deque ever got too long.
I think the limit should be to never have more than `connection_rate_limit_per_ip_count` items in there, but with our `-1` I am not sure about that.